tests/python/relax/test_frontend_nnef.py

```diff
@@ -0,0 +1,3142 @@
+    graph = get_case_graph(model_name)
+    for operation in graph.operations:
+        if operation.name == "variable":
+            tensor_name = operation.outputs["output"]
+
+            shape = operation.attribs["shape"]
+
+            assert (
+                operation.dtype == "scalar"
+            ), f"variable of type {operation.dtype} is not supported, please update verify_model"
+
+            data = np.ones(shape).astype("float32")
+
+            tensor = graph.tensors[tensor_name]
+            graph.tensors[tensor_name] = _nnef.Tensor(
+                tensor.name, tensor.dtype, shape, data, tensor.quantization
+            )
+
+    binding = {k: tvm.nd.array(v) for k, v in binding.items()}
+    expected = relax.transform.BindParams("main", binding)(expected)
+
+    mod = relax.frontend.nnef.from_nnef(graph)
+    tvm.ir.assert_structural_equal(mod, expected)
+
+
+def get_unary_mod(method, dt="float32", o_dtype=None):
+    global dtype
+    dtype = dt
+    if not o_dtype:
+        o_dtype = dtype
+
+    def _appl_shape(sh, osh=None):
+        global shape, o_shape
+        if not osh:
+            osh = sh
+        shape, o_shape = sh, osh
+
+        @tvm.script.ir.ir_module
+        class expected:
+            @R.function
+            def main(in1: R.Tensor(shape, dtype)) -> R.Tensor(o_shape, o_dtype):
+                R.func_attr({"num_input": 1})
+                with R.dataflow():
+                    lv: R.Tensor(o_shape, dtype=o_dtype) = method(in1)
+                    R.output(lv)
+                return lv
+
+        return expected
+
+    return _appl_shape
+
+
+def get_binary_mod(method, dt="float32", o_dtype=None):
+    global dtype
+    dtype = dt
+    if not o_dtype:
+        o_dtype = dtype
+
+    def _appl_shape(sh1, sh2=None, osh=None):
+        global shape1, shape2, o_shape
+        if not sh2:
+            sh2 = sh1
+        if not osh:
+            osh = sh1
+        shape1, shape2, o_shape = sh1, sh2, osh
+
+        @tvm.script.ir.ir_module
+        class expected:
+            @R.function
+            def main(
+                lhs: R.Tensor(shape1, dtype), rhs: R.Tensor(shape2, dtype)
+            ) -> R.Tensor(o_shape, o_dtype):
+                R.func_attr({"num_input": 2})
+                with R.dataflow():
+                    lv: R.Tensor(o_shape, dtype=o_dtype) = method(lhs, rhs)
+                    R.output(lv)
+                return lv
+
+        return expected
+
+    return _appl_shape
+
+
+# graph tests
+def test_copy():
+    @I.ir_module
+    class expected_2d:
+        @R.function
+        def main(input: R.Tensor((4, 16), dtype="float32")) -> R.Tensor((4, 16), dtype="float32"):
+            R.func_attr({"num_input": 1})
+            with R.dataflow():
+                lv = R.emit_te(topi.identity, input)
+                gv: R.Tensor((4, 16), dtype="float32") = lv
+                R.output(gv)
+            return gv
+
+    verify_model_struct("copy_2d", {}, expected_2d)
```

Comment:
i feel we can pass the nnef op, dimension, dtype as argument instead of a str like `copy_2d`. It could be like `verify_model_struct(op=nnef.copy, dim="2", dtype="float32",  {}, expected_2d)`, for some cases, we can also use relax blockbuilder to generate the expected IR as well

Reply:
Reworked it along the lines of your suggestion in https://github.com/apache/tvm/pull/17053/commits/c0d5b4408acb7edc811bd54af11094ba1838cc80